Cql/CqlSdkPrototype/Program.cs

```diff
@@ -0,0 +1,387 @@
+            {
+                Console.WriteLine($"Library already compiled: {elmVersionedIdentifier}");
+                continue;
+            }
+
+            var cSharpSourceCode = sourceCodePerName.Values.Single(); // We always expect a single source file
+            libraryCompilation = libraryCompilation with { CSharpSourceCode  = cSharpSourceCode, AssemblyBinary = assemblyBinary };
+            libraryCompilationsBuilder[elmVersionedIdentifier] = libraryCompilation;
+            Console.WriteLine($"Library compiled: {elmVersionedIdentifier}");
+            hasChanged = true;
+        }
+
+        return hasChanged
+            ? new ElmCompilation(this, libraryCompilations: libraryCompilationsBuilder.ToImmutable())
+            : this;
+    }
+
+    private static ServiceProvider BuildServiceProvider()
+    {
+        var serviceProvider =
+            new ServiceCollection()
+                .AddDebugLogging()
+                .AddCqlCodeGenerationServices()
+                .BuildServiceProvider(validateScopes: true);
+        return serviceProvider;
+    }
+
+    public IReadOnlyDictionary<string, ElmVersionedIdentifier> LibraryNames =>
+        _libraryCompilations
+            .ToDictionary(kv => kv.Key.Name.ToString(), kv => kv.Key);
+
+    public IReadOnlyDictionary<ElmVersionedIdentifier, string> SourceCodeByLibraryName =>
+        _libraryCompilations
+            .Where(kv => kv.Value.CSharpSourceCode is not null)
+            .ToDictionary(kv => kv.Key, kv => kv.Value.CSharpSourceCode!, ElmVersionedIdentifier.NameOnlyEqualityComparer);
+
+    public IReadOnlyDictionary<ElmVersionedIdentifier, byte[]> AssemblyBinariesByLibraryName =>
+        _libraryCompilations
+            .Where(kv => kv.Value.AssemblyBinary is not null)
+            .ToDictionary(kv => kv.Key, kv => kv.Value.AssemblyBinary!, ElmVersionedIdentifier.NameOnlyEqualityComparer);
+
+    public ElmCompilation SaveCSharpFilesToDirectory(DirectoryInfo directory)
+    {
+        if (!directory.Exists)
+            directory.Create();
+
+        foreach (var (libraryName, sourceCode) in SourceCodeByLibraryName)
+        {
+            var fileName = Path.Combine(directory.FullName, $"{libraryName}.cs");
+            File.WriteAllText(fileName, sourceCode);
+            Console.WriteLine($"Saved C# source code to file: {fileName}");
+        }
+
+        return this;
+    }
+
+    public ElmCompilation SaveAssemblyBinariesToDirectory(DirectoryInfo directory)
+    {
+        if (!directory.Exists)
+            directory.Create();
+
+        foreach (var (libraryName, assemblyBinary) in AssemblyBinariesByLibraryName)
+        {
+            var fileName = Path.Combine(directory.FullName, $"{libraryName}.dll");
+            File.WriteAllBytes(fileName, assemblyBinary);
+            Console.WriteLine($"Saved assembly binary to file: {fileName}");
+        }
+        return this;
+    }
+}
+
+
+public readonly record struct ElmVersionedIdentifier
```

Comment:
I think this struct is generic enough to be used in other contexts then just for the Elm, e.g. for the FHIR library resources etc. So what do you think about renaming this to just `VersionedIdentifier`?